crates/sui-faucet/src/faucet/simple_faucet.rs

```diff
@@ -548,6 +556,129 @@ impl SimpleFaucet {
+            };
+
+        let mut counter = 0;
+        let mut requests = Vec::new();
+        // We take the first 500 items off the queue or when it runs out.
+        while counter < 500 {
+            match timeout(RECV_TIMEOUT, req_consumer.recv()).await {
+                Ok(Some((uuid, address, amounts))) => {
+                    requests.push((uuid, address, amounts));
+                    counter += 1;
+                }
+
+                _ => {
+                    info!("Out of requests in queue.");
+                    break;
+                }
+            };
+        }
+
+        // Early exiting this function if there is nothing to batch.
+        if requests.is_empty() {
+            return Ok(TransactionDigest::ZERO);
+        }
+        let total_requests = requests.len();
+
+        let gas_cost = self.get_gas_cost().await?;
+        // The UUID here is for the batched request
+        let uuid = Uuid::new_v4();
+        info!(?uuid, "Batch transfer attemped");
+
+        // This loop is utilized to grab a coin that is large enough for the request
+        loop {
+            let gas_coin_response = self
+                .prepare_gas_coin(requests.len() as u64 + gas_cost, uuid)
+                .await;
+
+            match gas_coin_response {
+                GasCoinResponse::ValidGasCoin(coin_id) => {
+                    let tx_data = self
+                        .build_batch_pay_sui_txn(coin_id, requests, self.active_address, gas_cost)
+                        .await
+                        .map_err(FaucetError::internal)?;
+
+                    // Because we are batching transactions to faucet, we will just not use a real recipient for
+                    // sui address, and instead just fill it with the ZERO address.
+                    let recipient = SuiAddress::ZERO;
+                    {
+                        // Register the intention to send this transaction before we send it, so that if
+                        // faucet fails or we give up before we get a definite response, we have a
+                        // chance to retry later.
+                        let mut wal = self.wal.lock().await;
+                        wal.reserve(uuid, coin_id, recipient, tx_data.clone())
+                            .map_err(FaucetError::internal)?;
+                    }
+                    let response = self
+                        .sign_and_execute_txn(uuid, recipient, coin_id, tx_data)
+                        .await?;
+
+                    self.metrics
+                        .total_coin_requests_succeeded
+                        .add(total_requests as i64);
+                    // TODO (jian): add a check on the reqs and the coins sent
+                    return Ok(response.digest);
+                }
+
+                GasCoinResponse::UnknownGasCoin(coin_id) => {
+                    self.recycle_gas_coin(coin_id, uuid).await;
+                    return Err(FaucetError::FullnodeReadingError(format!(
```

Comment:
This seems like an intermittent fullnode issue.  If we fail at this point, will we ever retry the transaction?